Unified system for environment-based configuration

zetlen opened the pull request:
_This PR is a cherry-pick from #1198 to make that PR more digestible._ It needs to be merged before #1198.

## Description
Add tools in Buildpack to create, manage and validate configuration files for PWAs.

- Rename environment variables to fit a consistent UNDERSCORE_SEPARATED_NAMESPACE.
- Incorporate the brilliant new tool [camelspace](https://github.com/zetlen/camelspace) to abstract away the `process.env` and environment variables into configuration objects.
- Maintain a central `envVarDefinitions.json` in Buildpack for configuration across the project and all its tools:
    - Define the variables
    - Document the variables and their behavior
    - Validate according to simple types
    - Provide canonical defaults
- Create a `configureEnvironment` utility in Buildpack to replace the `validate-environment.js` script in Venia.
- Add functionality to announce deprecations of old environment variables, but continue to support them for some interval of time.

You can now translate easily between JS config objects and ENVIRONMENT_VARIABLE_OBJECTS, as well as drilling down to namespaces within them. Some tools (PWADevServer, upward-js) are already migrated to using these tools. Others are not quite yet.

## Related Issues
Closes #1116 

## Verification Steps
1. Run `yarn run build` to produce a working Buildpack CLI.
2. From repository root, run `node ./packages/pwa-buildpack/bin/buildpack create-env-file packages/venia-concept`
3. Observe the detailed logs and the created .env file.

## Proposed Labels for Change Type/Package
<!--- What type of change level would you suggest for this PR? -->
<!--- Major, Minor, or Patch? -->
<!--- See https://magento-research.github.io/pwa-studio/technologies/versioning/ for help -->
- [x] major (e.g x.0.0 - a breaking change)
- [ ] minor (e.g 0.x.0 - a backwards compatible addition)
- [ ] patch (e.g 0.0.x - a bug fix)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have updated the documentation accordingly, if necessary.
- [x] I have added tests to cover my changes, if necessary.

sirugh commented:
This is blocked until the following are merged.

#1215
#1216
#1217
#1219

jimbo commented on packages/pwa-buildpack/src/WebpackTools/PWADevServer.js:
```diff
@@ -100,79 +101,64 @@ const PWADevServer = {
+                if (!customOrigin) {
+                    console.warn(
+                        chalk.yellowBright(
+                            'Custom origins are enabled for this project, but one has not yet been set up. Run `npx @magento/pwa-buildpack init-custom-origin <projectRoot>` to initialize a custom origin.'
```
Consider suggesting a `yarn workspace` command.

zetlen replied:
That would only work if they were in this repository. If they're in their own, they need to run `npx` just to get a copy of buildpack. However, we could do `yarn buildpack`.

zetlen replied:
OK to merge it without the yarn-specific instructions for now?

jimbo commented on packages/pwa-buildpack/src/WebpackTools/PWADevServer.js:
```diff
@@ -35,27 +35,28 @@ const helpText = `
-                    !!parseInt(
-                        process.env.PWA_STUDIO_HOT_RELOAD_WITH_POLLING
-                    ) || false
+                    !!parseInt(devServerEnvVars.watchOptionsUsePolling) || false
```
Is the `|| false` is redundant here?

zetlen replied:
Potentially, but that option may be required to be boolean. Webpack does that sometimes.

jimbo commented on packages/pwa-buildpack/src/index.js:
```diff
@@ -1,4 +1,8 @@
+    ...Utilities,
+    ...WebpackTools,
+    Utilities,
+    WebpackTools
```
👍

jimbo commented on packages/upward-js/package.json:
```diff
@@ -33,6 +33,7 @@
   "dependencies": {
     "apollo-link": "~1.2.8",
     "apollo-link-http": "~1.5.11",
+    "camelspace": "^1.0.0",
```
```suggestion
    "camelspace": "~1.0.0",
```

zetlen commented on packages/pwa-buildpack/envVarDefinitions.json:
```diff
@@ -0,0 +1,256 @@
+        {
+            "name": "MAGENTO_BACKEND_URL",
+            "type": "exampleChanged",
+            "reason": "updates to our backend cloud instances",
```
This string gets used after the phrase "due to", so I think it should stay as it is.

zetlen commented on now-build.sh:
```diff
@@ -14,5 +14,4 @@ export NODE_ENV="production"
 yarn concurrently 'yarn workspace @magento/pwa-buildpack run build' 'yarn workspace @magento/peregrine run build:esm'
 # For the Venia build, any environment variables not set in now.json or other
 # scripts are filled in by copying the default .env file here.
-cp packages/venia-concept/.env.dist packages/venia-concept/.env
```
I remember that it was much faster to just copy the package.json files and do an NPM install, because then the installed packages are cached based on the package.json content alone. And then you can revert to the pre-installed stage before with Docker to speed up your install process, as long as you haven't changed the package files or yarn.lock.

TBH it would be easier if Dockerfiles let us copy globs.

sharkySharks replied:
sorry i missed your comment @jimbo , i don't know if I am missing something, but this is the now-build.sh file, this shouldn't have anything to do with the setup in the Dockerfile.

you might be remembering [this section](https://github.com/magento-research/pwa-studio/blob/develop/Dockerfile.dev#L16-L28) in the Dockerfile where we copy over the dependency files for install and then copy over the rest of the package file that won't change as much to take advantage of Docker cache.

zetlen commented on packages/pwa-buildpack/src/Utilities/configureEnvironment.js:
```diff
@@ -0,0 +1,277 @@
+        // the env isn't using the var with changes, no need to log
+        const isSet = env.hasOwnProperty(change.name);
+        switch (change.type) {
+            case 'exampleChanged':
```
May want to add defaultChanged back in here too.

zetlen commented:
Since @jimbo will be out for a few days, I'm dismissing his review, although he can feel free to violate parole and re-request changes. This is ready for review from @sirugh @tjwiebell, @dpatil-magento, @supernova-at, @sharkySharks optionally, and @jcalcaben since I have added a documentation draft.

sirugh commented:
@zetlen should this be rebased against `develop`?

zetlen commented:
@sirugh It should and I did. Good catch!

jcalcaben commented on pwa-devdocs/pwa-buildpack/configuration-management.md:
```diff
@@ -0,0 +1,187 @@
+# Configuration Management in Buildpack
```
This file should be an `index.md` file and placed inside a `configuration-management.md` directory.

Also, each sentence should be on a separate line to make it easier to review and add targeted comments.

Alternatively, you can move it into the `_drafts` folder and I can continue the work of getting it published on the docs site in a separate ticket.

zetlen replied:
Whoops, I'm sorry...I meant to put it in the drafts folder!

zetlen commented on packages/pwa-buildpack/envVarDefinitions.json:
```diff
@@ -0,0 +1,256 @@
+        },
+        {
+            "name": "SERVICE_WORKER_FILE_NAME",
+            "type": "removed",
```
I'm deprecating the document where most of them live.

zetlen commented on packages/pwa-buildpack/envVarDefinitions.json:
```diff
@@ -0,0 +1,256 @@
+            "reason": "The standard filename 'sw.js' is better for caching, stability, and compatibility, and a PWA Studio project should not register more than one ServiceWorker."
+        },
+        {
+            "name": "PWA_STUDIO_PUBLIC_PATH",
```
It does, and I want Docker users to start using the new variables according to their own needs, basically.

zetlen commented on packages/pwa-buildpack/src/Utilities/addImgOptMiddleware.js:
```diff
@@ -24,34 +24,26 @@ try {
-        redis: env.IMAGE_CACHE_REDIS_CLIENT
-    };
+function addImgOptMiddleware(app, config) {
+    const { backendUrl, cacheExpires, cacheDebug, redis } = config;
```
It does, good catch!

zetlen commented:
I'd like to put this on hold for the following reasons:

- I need to improve the docs
- I need to cherry-pick some of the CLI work, to complete the story for how to make `.env` files in the new system
- I need to add test coverage

Updated story above.

sirugh commented on babel.config.js:
```diff
@@ -1,96 +1,2 @@
-};
-
-module.exports = config;
+// Venia owns code style, and the rest of the project shares it.
```
Well... it will after this change :D Peregrine also uses this, so why not have this live in Peregrine? Maybe we should just keep this, prettier config, etc in a top-level `config/` directory and import them specifically within the sub-modules/packages.

zetlen replied:
Probably a better option is to make a babel preset. For now, I can switch this so it copies the babel config before publish.

zetlen replied:
@sirugh I reverted this change and I'll approach it from another angle in another PR.

sirugh commented on packages/peregrine/package.json:
```diff
@@ -1,18 +1,20 @@
 {
   "name": "@magento/peregrine",
-  "version": "2.1.0",
+  "version": "3.0.0-alpha.4",
```
Can you undo this version change?

zetlen replied:
Yep!

tjwiebell commented on packages/pwa-buildpack/src/Utilities/index.js:
```diff
@@ -1,4 +1,6 @@
     addImgOptMiddleware: require('./addImgOptMiddleware'),
-    configureHost: require('./configureHost')
+    configureEnvironment: require('./configureEnvironment'),
+    configureHost: require('./configureHost'),
```
Jimmy has to take his adoration away now; lets alphabetize, keep him happy.

tjwiebell commented on packages/pwa-buildpack/src/Utilities/loadEnvironment.js:
```diff
@@ -0,0 +1,299 @@
+/**
+ * Replace the envalid default reporter, which crashes the process, with an
+ * error-returning reporter that a consumer can handle.
+ *
```
:point_up:

zetlen replied:
What...did you mean by this?

tjwiebell replied:
Remove extra line break, you got it.

tjwiebell commented on packages/pwa-buildpack/src/WebpackTools/plugins/ServiceWorkerPlugin.js:
```diff
@@ -3,10 +3,12 @@ const WorkboxPlugin = require('workbox-webpack-plugin');
 const WriteFileWebpackPlugin = require('write-file-webpack-plugin');
 const optionsValidator = require('../../util/options-validator');
 
+// No longer modifiable since v3
```
This will probably be v4 now

zetlen commented on packages/pwa-buildpack/package.json:
```diff
@@ -34,19 +38,21 @@
     "@magento/express-sharp": "~2.1.1-newdeps.1",
-    "@magento/upward-js": "~2.1.0",
-    "ajv": "~6.9.2",
+    "@magento/upward-js": "2.1.0",
```
No, that was a mistake. Fixed!

zetlen commented on now-build.sh:
```diff
@@ -14,5 +14,4 @@ export NODE_ENV="production"
 yarn concurrently 'yarn workspace @magento/pwa-buildpack run build' 'yarn workspace @magento/peregrine run build:esm'
 # For the Venia build, any environment variables not set in now.json or other
 # scripts are filled in by copying the default .env file here.
-cp packages/venia-concept/.env.dist packages/venia-concept/.env
```
[Replaced them!](https://github.com/magento-research/pwa-studio/pull/1224/files#diff-d6f6fe8b451de4ac0dc51407b5b74a26R16)

zetlen commented on now-build.sh:
```diff
@@ -12,7 +12,7 @@ export NODE_ENV="production"
-cp packages/venia-concept/.env.dist packages/venia-concept/.env
+
+# No .env file generation is necessary, since the only required variable without
+# a default, MAGENTO_BACKEND_URL, is set in the build environment by now.json.
```
It actually is stored in `envVarDefinitions` now. It wasn't working because I messed up pushing the environment into the Webpack EnvironmentPlugin. I fixed it and the deployed site has working payments again. Thanks for catching this.

zetlen commented:
I believe that now.sh deployment failed because of a babel cache. I redeployed from my desktop and https://venia-4wssjviar.now.sh/ works.

jcalcaben commented on packages/pwa-buildpack/envVarDefinitions.json:
```diff
@@ -0,0 +1,256 @@
+                },
+                {
+                    "name": "CUSTOM_ORIGIN_SUBDOMAIN",
+                    "type": "str",
```
Can we use "string" instead of "str" in this file? It reads easier.

zetlen replied:
From another comment:
>It would, but that field is used as an [`envalid` validator type](https://www.npmjs.com/package/envalid#validator-types), so we're stuck with their name scheme.

jcalcaben commented on packages/pwa-buildpack/envVarDefinitions.json:
```diff
@@ -0,0 +1,256 @@
+                },
+                {
+                    "name": "DEV_SERVER_PORT",
+                    "type": "num",
```
"number" would read easier than "num"

zetlen replied:
It would, but that field is used as an [`envalid` validator type](https://www.npmjs.com/package/envalid#validator-types), so we're stuck with their name scheme.

sirugh replied:
@zetlen is there a possibility we can do a remap of these somewhere as an intermediate step? I actually had the same feedback as @jcalcaben but I did see that it was the types used by `envalid` and ended up deleting the comment.

dpatil-magento commented:
Regression on Venia looks good. But @zetlen - We tried to follow instruction of devdocs but not working as expected.

1. Cloned fresh repo > checkout this branch > `yarn install` > ' yarn build'  errors - 
 ⚠  No .env file in .

     You can autogenerate an .env file in your PWA project root by running the command
     npx @magento/pwa-buildpack create-env-file ..
  ⓧ  Missing required environment variables:
     MAGENTO_BACKEND_URL: Connect to an instance of Magento 2.3 by specifying its
     public domain name. (eg. "https://magento.localhost/")
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed.
Exit code: 1

2. After yarn build by cp .env manually if we run > `npx @magento/pwa-buildpack buildpack` or `npx @magento/pwa-buildpack`  from pwa-studio root directory getting error "command not found: pwa-buildpack"

jcalcaben commented on pwa-devdocs/_drafts/configuration-management/index.md:
```diff
@@ -0,0 +1,431 @@
+
+## Command Line Interface
+
+The `buildpack` command is a toolkit with subcommands.
```
```suggestion
The `buildpack` command is a toolkit with subcommands available in the `pwa-buildpack` module.

When you have the module added as a dependency in your project, you can run it in the command line: `npx buildpack`
```

jcalcaben commented on pwa-devdocs/_drafts/configuration-management/index.md:
```diff
@@ -0,0 +1,431 @@
+
+_Note: It would be unusual to use the above example directly; the more common use of `buildpack load-env` is in NPM scripts as described above._
+
+### `buildpack create-custom-origin <dir>`
```
The `<dir>` argument is pretty straightforward for the other two commands(create in this directory or load from this directory), but what is the significance of `<div>` for this command?

zetlen replied:
Buildpack creates the custom origin subdomain name based on a hash of the directory path. In general, we want to explicitly pass directory references to Buildpack commands because so many of them are reliant on a PWA directory for context. In this case it's especially relevant because the custom origin hash will change depending on the string path of the directory you're in.

jcalcaben commented on pwa-devdocs/_drafts/configuration-management/index.md:
```diff
@@ -0,0 +1,431 @@
+_Note: It is not recommended to globally install buildpack with `yarn global add`._
+_Individual projects should use their own versions and a globally installed buildpack can cause unexpected behavior._
+
+### `buildpack create-env-file <dir>`
```
Would be worth noting that this command expects the `MAGENTO_BACKEND_URL` environment variable to be set prior or it will create a `.env` file with a blank backend url entry.
```sh
MAGENTO_BACKEND_URL="https://magento.test/" npx buildpack create-env-file .
```

zetlen replied:
True, we could add that there. It's not a cross-platform command (it won't work on Windows) but we could have a Windows version with `SET`.

zetlen commented:
This fixes #85 to a great extent.

zetlen commented:
@magento-research/core-pwa-devs I've changed bits of this PR enough that we need another brief pass of code review. The main changes since the last code review are:

- A couple of new flags in the CLI to make cloning and starting the repo easier, specifically `load-env --core-dev-mode` and `create-env-file --use-examples`
- Removing all the different example backends and replacing them with one consistent one, a Cloud Starter environment with Fastly enabled
- A few changes to how the dotenv file is generated

zetlen commented on docker/.env.docker.dev:
```diff
@@ -9,7 +9,7 @@ PWA_STUDIO_PUBLIC_PATH=pwa-docker.localhost
 DEV_SERVER_SERVICE_WORKER_ENABLED=0
 DEV_SERVER_WATCH_OPTIONS_USE_POLLING=0
-MAGENTO_BACKEND_URL=https://release-dev-231-npzdaky-zddsyhrdimyra.us-4.magentosite.cloud/
+MAGENTO_BACKEND_URL=https://master-7rqtwti-mfwmkrjfqvbjk.us-4.magentosite.cloud/
```
That's wise advice, but we also don't want to publicize and therefore have to maintain too many instances. @dpatil-magento should weigh in here about how to balance those needs.
